src/man/man3/AvoidObstacle.3
.TH AvoidObstacle 3 "29 May 2002" "HulkRobot" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AvoidObstacle \- 
.SH SYNOPSIS
.br
.PP
\fC#include <AvoidObstacle.h>\fR
.PP
.SS Public Methods

.in +1c
.ti -1c
.RI "\fBAvoidObstacle\fR (int priority, int samplingPeriod, int stackSize, \fBgyro\fR *aG, \fBusonic\fR *aU, \fBwheel\fR *aW)"
.br
.ti -1c
.RI "\fB~AvoidObstacle\fR ()"
.br
.ti -1c
.RI "int \fBinit\fR (void)"
.br
.ti -1c
.RI "int \fBstep\fR (void)"
.br
.ti -1c
.RI "int \fBrun\fR (void)"
.br
.ti -1c
.RI "int \fBstop\fR (void)"
.br
.in -1c
.SH CONSTRUCTOR & DESTRUCTOR DOCUMENTATION
.PP 
.SS AvoidObstacle::AvoidObstacle (int _priority, int _samplingPeriod, int stackSize, \fBgyro\fR * aG, \fBusonic\fR * aU, \fBwheel\fR * aW)
.PP
Definition at line 18 of file AvoidObstacle.cc.
.PP
.nf
00020 {
00021 
00022   Error( ErrorDebug, "AvoidObstacle initializing");
00023   priority        = _priority;
00024   samplingPeriod= _samplingPeriod;
00025   delayTicks      = samplingPeriod/OSTickPeriod;
00026 
00027   taskMode        = TaskStopped;
00028   taskSem = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00029   tickerSem       = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00030   taskExitSem     = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00031   tickerExitSem   = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00032 
00033   aGyro = aG;
00034   aUSonic = aU;
00035   aWheel = aW;
00036 
00037   Assert( taskSem && taskExitSem && tickerSem && tickerExitSem );
00038 
00039   taskId  = taskSpawn("AvoidObstacle", priority, VX_FP_TASK, 
00040                           stackSize, (FUNCPTR) taskModuleStub,
00041                           (int) this, 0, 0, 0, 0, 0, 0, 0, 0, 0);
00042   Assert(taskId!=ERROR);
00043 
00044   semTake(taskExitSem, WAIT_FOREVER);
00045   
00046   tickerTaskId    = taskSpawn("AvoidObstacle", TickerTaskPriority, 0, 
00047                           TickerTaskStackSize,
00048                           (FUNCPTR) taskModuleTickerStub,
00049                           (int) this,
00050                           0, 0, 0, 0,
00051                           0, 0, 0, 0, 0);
00052   Assert(tickerTaskId!=ERROR);
00053   semTake(tickerExitSem, WAIT_FOREVER);
00054 
.fi
.SS AvoidObstacle::~AvoidObstacle ()
.PP
Definition at line 56 of file AvoidObstacle.cc.
.PP
.nf
00057 {
00058   taskMode=TaskTerminate;
00059   semGive(tickerSem);
00060   semGive(taskSem);
00061   semTake(tickerExitSem, WAIT_FOREVER);
00062   semTake(taskExitSem, WAIT_FOREVER);
00063   semDelete(taskSem);
00064   semDelete(taskExitSem);
00065   semDelete(tickerSem);
00066   semDelete(tickerExitSem);
00067 
00068   Error( ErrorDebug, "AvoidObstacle destroyed");
00069 
.fi
.SH MEMBER FUNCTION DOCUMENTATION
.PP 
.SS int AvoidObstacle::init (void)
.PP
Definition at line 229 of file AvoidObstacle.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00230 {
00231 
.fi
.SS int AvoidObstacle::run (void)
.PP
Definition at line 240 of file AvoidObstacle.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00241 {
00242   taskMode=TaskRunning;
00243   semGive(tickerSem);
00244   return 0;
00245 
.fi
.SS int AvoidObstacle::step (void)
.PP
Definition at line 233 of file AvoidObstacle.cc.
.PP
.nf
00234 {
00235   taskMode=TaskStep;
00236   semGive(taskSem);
00237   return 0;
00238 
.fi
.SS int AvoidObstacle::stop (void)
.PP
Definition at line 247 of file AvoidObstacle.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00248 {
00249   taskMode=TaskStopped;
00250   return 0;
00251 
.fi


.SH AUTHOR
.PP 
Generated automatically by Doxygen for HulkRobot from the source code.

src/bbc/AvoidObstacle.cc
//
//	File AvoidObstacle.cc
//
//	Avoid obstacle Behavior Controller Module
//
//	Framework is originally written by Yunho Jeon
//
//	Written by Insop Song
//	1997 12 1
//	ver 0.1
//

#include "dfl.h"
#include "AvoidObstacle.h"

unsigned AvoidObstacle::range[16];

AvoidObstacle::AvoidObstacle(int _priority, int _samplingPeriod, 
		int stackSize, gyro *aG, usonic *aU, wheel *aW)
{

  Error( ErrorDebug, "AvoidObstacle initializing");
  priority 	= _priority;
  samplingPeriod= _samplingPeriod;
  delayTicks	= samplingPeriod/OSTickPeriod;

  taskMode	= TaskStopped;
  taskSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  tickerSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  taskExitSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  tickerExitSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);

  aGyro = aG;
  aUSonic = aU;
  aWheel = aW;

  Assert( taskSem && taskExitSem && tickerSem && tickerExitSem );

  taskId 	= taskSpawn("AvoidObstacle", priority, VX_FP_TASK, 
				stackSize, (FUNCPTR) taskModuleStub,
				(int) this, 0, 0, 0, 0, 0, 0, 0, 0, 0);
  Assert(taskId!=ERROR);

  semTake(taskExitSem, WAIT_FOREVER);
  
  tickerTaskId	= taskSpawn("AvoidObstacle", TickerTaskPriority, 0, 
				TickerTaskStackSize,
				(FUNCPTR) taskModuleTickerStub,
				(int) this,
				0, 0, 0, 0,
				0, 0, 0, 0, 0);
  Assert(tickerTaskId!=ERROR);
  semTake(tickerExitSem, WAIT_FOREVER);
}

AvoidObstacle::~AvoidObstacle()
{
  taskMode=TaskTerminate;
  semGive(tickerSem);
  semGive(taskSem);
  semTake(tickerExitSem, WAIT_FOREVER);
  semTake(taskExitSem, WAIT_FOREVER);
  semDelete(taskSem);
  semDelete(taskExitSem);
  semDelete(tickerSem);
  semDelete(tickerExitSem);

  Error( ErrorDebug, "AvoidObstacle destroyed");
}

//
// C stub functions to call C++ member functions from
// a newly spawned task
//

void AvoidObstacle::taskModuleStub(AvoidObstacle *This, int d2, int d3, 
					int d4, int d5, int d6, int d7, 
					int d8, int d9, int d10)
{
  Error( ErrorDebug, "AvoidObstacle taskModuleStub ");
  This->task();
}

void AvoidObstacle::taskModuleTickerStub(AvoidObstacle *This, int d2, 
			      int d3, int d4, int d5, int d6, int d7, 
			      int d8, int d9, int d10)
{
  Error( ErrorDebug, "AvoidObstacle taskModuleTickerStub");
  This->ticker();
}

void AvoidObstacle::ticker(void)
{
  semGive(tickerExitSem);
  while(taskMode!=TaskTerminate) {

    if(taskMode==TaskStopped || taskMode==TaskPaused) 
       semTake(tickerSem, WAIT_FOREVER);

    while(taskMode==TaskRunning) {
      Error( ErrorNone, "ticker::semGive" );
      semGive(taskSem);
      taskDelay(delayTicks);
    }
  }
  semGive(tickerExitSem);
  // task exit - deleted automatically
}

void AvoidObstacle::task(void)
{
  semGive(taskExitSem); 		// signal that it has been created
  while(taskMode!=TaskTerminate) {
    semTake(taskSem, WAIT_FOREVER);	// wait until ticker ticks
    if (taskMode!=TaskRunning && taskMode!=TaskStep) continue;
    Error( ErrorNone, "task tick" );

    // task code HERE
    for(int i=0;i<16;i++) {
      range[i] = aUSonic->usData(i);
    }
    //printf(" [15] %d  [0], %d  [1],%d\n",range[15],range[0],range[1]);

    //
    // check front 3 sensors , configuration  [15]-[0]-[1]
    //

    if((range[15] > DangerZone) && (range[0] > DangerZone)
			      && (range[1] > DangerZone)){
      turnMode = GO;
      Error( ErrorDebug, "Go" );
    } 
    /*
    else if(!(range[15]>DangerZone) && (range[0]>DangerZone) 
      && (range[1]>DangerZone)){
      turnMode = TURN_SOFT_RIGHT;
      Error( ErrorNone, "TURN_SOFT_LEFT" );
    }
    else if((range[15]>DangerZone) && (range[0]>DangerZone) 
      && !(range[1]>DangerZone)){
      turnMode = TURN_SOFT_RIGHT;
      Error( ErrorNone, "TURN_SOFT_RIGHT" );
    }
    */
    else if((!(range[15]>DangerZone) || !(range[0]>DangerZone)) 						/* && (range[1]>DangerZone)*/ ){
      turnMode = TURN_HARD_RIGHT;
      Error( ErrorNone, "TURN_HARD_RIGHT" );
    }
    else if(/* (range[15]>DangerZone) && */ (!(range[0]>DangerZone) 
				 || !(range[1]>DangerZone)) ){
      turnMode = TURN_HARD_LEFT;
      Error( ErrorNone, "TURN_HARD_LEFT" );
    }


/*
 * for stop snd back condition 
 *
    if((!(range[15] > StopZone) || !(range[0] > StopZone)
         || !(range[1] > StopZone)) && !(turnMode == STOP)	
				     &&	!(turnMode == BACK)){
      turnMode = STOP;
      Error( ErrorNone, "STOP" );
    } 
    if((!(range[15] > DangerZone) || !(range[0] > DangerZone)
	      || !(range[1] > DangerZone)) && (turnMode==STOP)){
      turnMode = BACK;
      Error( ErrorNone, "BACK" );
    } 

 *
 *
 */



    switch(turnMode) {
      case(GO):
	aWheel->setSpeed(0.1);
	aWheel->setTurnSpeed(0.0);
	break;
      case(TURN_SOFT_RIGHT):
	aWheel->setSpeed(0.05);
	aWheel->setTurnSpeed(0.05);
	break;
      case(TURN_SOFT_LEFT):
	aWheel->setSpeed(0.05);
	aWheel->setTurnSpeed(-0.05);
	break;
      case(TURN_HARD_RIGHT):
	aWheel->setSpeed(0.01);
	aWheel->setTurnSpeed(0.05);
	break;
      case(TURN_HARD_LEFT):
	aWheel->setSpeed(0.01);
	aWheel->setTurnSpeed(-0.05);
	break;
      case(STOP):
	aWheel->setSpeed(0.0);
	aWheel->setTurnSpeed(0.0);
	break;
      case(BACK):
	aWheel->setSpeed(-0.05);
	aWheel->setTurnSpeed(0.0);
	break;
    }
    // task code till HERE

  }
	aWheel->setSpeed(0.0);
	aWheel->setTurnSpeed(0.0);

  semGive(taskExitSem); 
  // task exit - deleted automatically
}

unsigned AvoidObstacle::minimum(unsigned a,unsigned b,unsigned c)
{
  unsigned min = a;
  if(min>b)
    min = b;
  if(min>c);
    min = c;

  return c;
}
  
    
int AvoidObstacle::init(void)
{
}

int AvoidObstacle::step(void)
{
  taskMode=TaskStep;
  semGive(taskSem);
  return 0;
}

int AvoidObstacle::run(void)
{
  taskMode=TaskRunning;
  semGive(tickerSem);
  return 0;
}

int AvoidObstacle::stop(void)
{
  taskMode=TaskStopped;
  return 0;
}

